assets/app/scripts/services/apiEndpoint_v2.js

```diff
@@ -0,0 +1,311 @@
+       }
+
+       return normalized;
+    };
+
+    // port of restmapper.go#kindToResource
+    var kindToResource = function(kind) {
+      if (!kind) {
+        return "";
+      }
+      var resource = String(kind).toLowerCase();
+      if (resource.endsWith('status')) {
+        resource = resource + 'es';
+      } else if (resource.endsWith('s')) {
+        // no-op
+      } else if (resource.endsWith('y')) {
+        resource = resource.substring(0, resource.length-1) + 'ies';
+      } else {
+        resource = resource + 's';
+      }
+      // make sure it is a known resource
+      if (!apiExistsFor(resource)) {
+        Logger.warn('Unknown resource "' + resource + '"');
+        return undefined;
+      }
+      return resource;
+    };
+
+    // TODO: this is not efficient. make another map & do a lookup,
+    // also use apiVersion (if we need it)
+    var apiExistsFor = function(resource, apiVersion) {
+      var found = false;
+      _.each(apis, function(api) {
+        _.each(api, function(version) {
+          _.each(version, function(res, key) {
+            if(resource === key) {
+              found = true;
+            }
+          });
+        });
+      });
+      return found;
+    };
+
+    // deprecated?
+    var ensureVersion = function(api, preferredAPIVersion) {
+      return preferredAPIVersion || (api && api.defaultVersion);
+    };
+
+    // Bother, URITemplate has no ability to do conditional {+group}, so
+    // templates have to be more complex
+    var API_TEMPLATE = "{protocol}://{+hostPort}{+prefix}/{version}/";
+    var API_GROUP_TEMPLATE = "{protocol}://{+hostPort}{+prefix}/{+group}/{version}/";
+
+    var URL_GET_LIST              = "{resource}{?q*}";
+    var URL_OBJECT                = "{resource}/{name}{/subresource*}{?q*}";
+    var URL_NAMESPACED_GET_LIST   = "namespaces/{namespace}/{resource}{?q*}";
+    var URL_NAMESPACED_OBJECT     = "namespaces/{namespace}/{resource}/{name}{/subresource*}{?q*}";
+
+    var findTemplateFor = function(name, namespace, group) {
+      var base = group ?
+                  API_GROUP_TEMPLATE :
+                  API_TEMPLATE;
+      var rest = name ?
+                  namespace ?
+                    URL_NAMESPACED_OBJECT :
+                    URL_OBJECT :
+                  namespace ?
+                    URL_NAMESPACED_GET_LIST :
+                    URL_GET_LIST;
+      return (base + rest);
+    };
+
+    var protocol = function(isWebsocket) {
+      return isWebsocket ?
+              window.location.protocol === "http:" ?
+                "ws" :
+                "wss" :
+              window.location.protocol === "http:" ?
+                "http" :
+                "https";
+    };
+
+    var cleanCopyParams = function(params) {
+      params = (params &&  angular.copy(params)) || {};
+      delete params.namespace;
+      return params;
+    };
+
+    var findNamespace = function(context, params) {
```

Comment:
Let me see if I can make that more readable. Its a deliberate functional style thing (the best functions return once and return immediately), but I'm sure I can find a way to clean it up.